src/connection.ts

```diff
@@ -1774,6 +1777,442 @@ class Connection extends EventEmitter {
     }
 
     this.transitionTo(this.STATE.CONNECTING);
+
+    const controller = new AbortController();
+
+    // Abort the connect operation if we reached the connection timeout
+    const connectTimer = setTimeout(() => {
+      controller.abort();
+    }, this.config.options.connectTimeout);
+
+    // Abort the connect operation if the connection is closed
+    const onClose = () => { controller.abort(); };
+    const closeSignal = this.closeController.signal;
+    closeSignal.addEventListener('abort', onClose, { once: true });
+
+    this.establishConnection(controller.signal).finally(() => {
+      clearTimeout(connectTimer);
+      closeSignal.removeEventListener('abort', onClose);
+    }).catch((err) => {
+      if (!closeSignal.aborted && err instanceof Error && err.name === 'AbortError') {
+        const message = `Failed to connect to ${this.config.server}${this.config.options.port ? `:${this.config.options.port}` : `\\${this.config.options.instanceName}`} in ${this.config.options.connectTimeout}ms`;
+        this.debug.log(message);
+
+        err = new ConnectionError(message, 'ETIMEOUT');
+      }
+
+      process.nextTick(() => {
+        this.emit('connect', err);
+        this.transitionTo(this.STATE.FINAL);
+      });
+    });
+  }
+
+  /**
+   * @private
+   */
+  async lookupInstancePort(signal: AbortSignal) {
+    const lookup = new InstanceLookup();
+
+    return (await promisify(lookup.instanceLookup.bind(lookup))({
+      server: this.config.server,
+      instanceName: this.config.options.instanceName!,
+      timeout: this.config.options.connectTimeout,
+      signal: signal
+    }))!;
+  }
```

Comment:
The instance lookup code itself should be refactored to use promises. See https://github.com/tediousjs/tedious/pull/1211 for some previous work that was never finished.